Userland/xargs.cpp

```diff
@@ -0,0 +1,307 @@
+                total_command_length = 0;
+                return Unget;
+            } else {
+                child_argv.append(strndup(item.characters_without_null_termination(), item.length()));
+                ++items_used_for_this_command;
+            }
+        }
+
+        return Continue;
+    });
+
+    if (!fail && !child_argv.is_empty())
+        fail = !run_command(child_argv, verbose, is_stdin, devnull_fd);
+
+    if (!is_stdin)
+        fclose(fp);
+
+    return fail ? 1 : 0;
+}
+
+bool read_items(FILE* fp, char entry_separator, Function<Decision(StringView)> callback)
+{
+    bool fail = false;
+
+    for (;;) {
+        char* item = nullptr;
+        size_t buffer_size = 0;
+
+        auto item_size = getdelim(&item, &buffer_size, entry_separator, fp);
+
+        if (item_size < 0) {
+            // getdelim() will return -1 and set errno to 0 on EOF.
+            if (errno != 0) {
+                perror("getdelim");
+                fail = true;
+            }
+            break;
+        }
+
+        Decision decision;
+        do {
+            decision = callback(item);
+            if (decision == Stop) {
+                free(item);
+                return true;
+            }
+        } while (decision == Unget);
+
+        free(item);
+    }
+
+    return fail;
+}
+
+bool run_command(Vector<char*>& child_argv, bool verbose, bool is_stdin, int devnull_fd)
+{
+    child_argv.append(nullptr);
+
+    if (verbose) {
+        StringBuilder builder;
+        builder.join(" ", child_argv);
+        fprintf(stderr, "xargs: %s\n", builder.to_string().characters());
+        fflush(stderr);
+    }
+
+    auto pid = fork();
+    if (pid < 0) {
+        perror("fork");
+        return false;
+    }
+
+    if (pid == 0) {
+        if (is_stdin)
+            dup2(devnull_fd, STDIN_FILENO);
+
+        execvp(child_argv[0], child_argv.data());
+        exit(1);
+    }
+
+    for (auto* ptr : child_argv)
+        free(ptr);
+
+    child_argv.clear();
+
+    int wstatus = 0;
+    if (waitpid(pid, &wstatus, 0) < 0) {
+        perror("waitpid");
+        return false;
+    }
+
+    if (WIFEXITED(wstatus)) {
+        if (WEXITSTATUS(wstatus) != 0)
+            return false;
+    } else {
+        return false;
+    }
+
+    return true;
+}
+
+ParsedInitialArguments::ParsedInitialArguments(Vector<const char*>& arguments, const StringView& placeholder)
+{
+    m_all_parts.ensure_capacity(arguments.size());
+    bool some_argument_has_replacement = false;
+
+    for (auto argument : arguments) {
+        StringView arg { argument };
+        Vector<StringView> parts;
+        size_t previous_index = 0;
+        if (!placeholder.is_empty()) {
+            // Split the argument into segments delimited by the placeholder.
+            if (auto first_index = arg.find_first_of(placeholder); first_index.has_value()) {
+                some_argument_has_replacement = true;
+                while (first_index.has_value()) {
```

Comment:
Why does this have to be nested (`while` inside a `for`)?

Reply:
Wait, that wasn't a `for`, it was an `if`!

But still, I don't think it has to be nested.

Reply:
wdym?
"loop over the inner segments, then add the last bit" but that only applies if there's a placeholder in there somewhere

un-nesting this would just add another if at the end, not sure what we gain here.
the if with the initialiser might be a bit weird, I'll move that variable out.

but I see no real point to un-nesting that while, if anything, it hurts readability here

Reply:
* Add all but last (if any)
* Add the last

reads cleaner to me then

* If there are more than one:
    * Add all but last
    * Add the last
* Else:
    * Add the only one

Don't you agree? And I don't see why you'd need an `if` at the end?

Reply:
we can't find the last in a clean way here.
"last" is offset `placeholder.length()` bytes if the placeholder exists, and not otherwise

Reply:
No, just keep an offset where to start the next part from. Initially it's 0, if you find the placeholder it becomes the end of that placeholder. In other words, that "offset `placeholder.length()` bytes" should happen at the previous iteration of the loop.

Reply:
Alright, let's try that.

Reply:
Fixing the logic takes a minute, renaming things right takes 19 minutes.
let's just revert back to `var0`, `var1`, etc

Reply:
No worries, naming things is one of the hardest computer science problems :smile: